src/librustdoc/clean/utils.rs

```diff
@@ -624,3 +624,24 @@ where
     *cx.impl_trait_bounds.borrow_mut() = old_bounds;
     r
 }
+
+crate fn find_closest_parent_module(tcx: TyCtxt<'_>, def_id: DefId) -> Option<DefId> {
+    if item.is_fake() {
+        // FIXME: is this correct?
+        None
+    // If we're documenting the crate root itself, it has no parent. Use the root instead.
+    } else if item.def_id.is_top_level_module() {
+        Some(item.def_id)
+    } else {
+        let mut current = def_id;
+        // The immediate parent might not always be a module.
+        // Find the first parent which is.
+        while let Some(parent) = tcx.parent(current) {
+            if tcx.def_kind(parent) == DefKind::Mod {
+                return Some(parent);
+            }
+            current = parent;
+        }
+        None
+    }
+}
```

Comment:
It seemed to work fine without these top-level conditionals (all the tests passed). I'm not sure to get this to work for all use cases, because sometimes we don't have an `Item` (e.g. with macros). Should we just remove this conditional logic (and leave it where it was, in the intra-doc link code)?

Reply:
Actually, `is_fake()`'s logic only needs a `DefId`, so I could still get this to work (if I extract out `is_fake()` so it operates on `DefId`s)
.

Reply:
Oh, I realized why this isn't necessary: you can't give a visibility to either the crate root or blanket impls.

I don't feel strongly about whether you handle fake IDs here or in intra-doc-links, but if you remove it from here please leave a comment saying it will ICE if you use it on a fake id.

Reply:
Hmm, why will it panic on a fake ID?

Reply:
It's used as an index into the crate metadata, it will panic with an index out of bounds error. Fake IDs weren't serialized to disk, so they aren't present in the metadata; really they should be a separate type that's only in rustdoc.

Reply:
Huh, okay. I'll add a comment.

Reply:
Done.